### PR TITLE
Add support for translations

### PR DESCRIPTION
**Pull request by avivace**
Closes https://github.com/gbdev/gb-asm-tutorial/issues/25

Based on the i18n support implemented in https://github.com/google/comprehensive-rust

- [x] Credit https://github.com/google/comprehensive-rust for the translation system add headers+license in i18n related code
- [x] Add TRANSLATING file explaining how to contribute to translations and add new ones
- [x] Add a sample translation to start with (e.g. one page in Italian from https://github.com/gbdev/gb-asm-tutorial/pull/32
- [x] Mention the translation guidelines in the README and in the book
- [x] Adapt the current build pipeline to go through the gettext preprocessor 
- [x] Add a submenu linking to available translated versions of the book
- [x] Update CI to build translated versions of the book
- [x] If all of this is done, rebase it in two commits: one for adding the i18n code and the other for the italian po

### CODE REVIEW

File: .github/workflows/build-and-deploy.yml [hunk 1]
```diff
@@ -44,6 +49,17 @@ jobs:
...
+    - name: Build IT translation
+      working-directory: gb-asm-tutorial/
+      env:
+        MDBOOK_BUILD__CREATE_MISSING: "false" # Prevent creating missing files in SUMMARY.md
+        MDBOOK_OUTPUT__LINKCHECK__COMMAND: "../mdbook-linkcheck"
+        MDBOOK_OUTPUT__LINKCHECK__OPTIONAL: "false"
+        MDBOOK_BOOK__LANGUAGE: "it"
+      run: |
+        mdbook build -d book/it
+        mv book/it/custom book/custom/it
```
Comment (ISSOtm) [inline .github/workflows/build-and-deploy.yml:61]:
```suggestion
        mdbook build
        for f in po/*.po; do
            lang="${f%.po}"
            MDBOOK_BOOK__LANGUAGE="$lang" mdbook build -d book/"$lang"
            mv book/"$lang"/custom book/custom/"$lang"
        done
```
This allows adding new languages without touching the CI workflow, and is also a bit more DRY.

File: .gitignore [hunk 2]
```diff
@@ -1,3 +1,4 @@
 /book/
 /target/
 .DS_Store
+po/messages.pot
```
Comment (ISSOtm) [inline .gitignore:4]:
```suggestion
/po/messages.pot
```
I suppose?

File: README.md [hunk 3]
```diff
@@ -10,13 +10,19 @@ To deploy gb-asm-tutorial locally:
 
 1. Install [Rust](https://www.rust-lang.org/tools/install) and [mdBook](https://github.com/rust-lang/mdBook#readme).
   mdBook powers the book itself, Rust is used for some custom plugins.
+```
+$ cargo install mdbook
+$ cargo install --path i18n-helpers
+```
```
Comment (ISSOtm) [inline README.md:16]: Two spaces should be added before each of these lines, otherwise the code block breaks the list.

File: README.md [hunk 4]
```diff
@@ -10,13 +10,19 @@ To deploy gb-asm-tutorial locally:
...
 2. Within a terminal pointed at the directory `book.toml` is in, run mdBook (`mdbook build` / `mdbook watch` / `mdbook serve`).
 3. The HTML files are in `book/custom/`.
 
   ⚠️ `book/html/` contains only partially processed files.
   This folder is what gets served when running `mdbook serve`, so you may see some custom markup missing if using that.
   As a workaround, you can manually open the files in the `book/custom/` folder in your browser, they just won't auto-refresh on changes.
 
+To add a translation or contribute on an existing one, please see [TRANSLATING](TRANSLATING.md).
```
Comment (ISSOtm) [inline README.md:24]: What about moving this just below the "Contributing" header? This would increase its visibility.

File: README.md [hunk 5]
```diff
@@ -55,4 +61,6 @@ Different parts of gb-asm-tutorial are subject to different licenses:
 
 - All the code contained within the tutorial itself is licensed under <a rel="license" href="http://creativecommons.org/publicdomain/zero/1.0/">CC0</a>. *To the extent possible under law, all copyright and related or neighboring rights to code presented within GB ASM Tutorial have been waived. This work is published from France.*
 - The contents (prose, images, etc.) of this tutorial are licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>.
-- Code used to display and format the site is licensed under the [MIT License](https://github.com/ISSOtm/gb-asm-tutorial/blob/master/LICENSE) unless otherwise specified.
+- Code used to display and format the site is licensed under the [MIT License](https://github.com/gbdev/gb-asm-tutorial/blob/master/LICENSE) unless otherwise specified.
+- The code related to the i18n support is originally from Google's [Comprehensive Rust](https://github.com/google/comprehensive-rust) and it's released under the [Apache License 2.0](https://github.com/gbdev/gb-asm-tutorial/blob/master/i18n-helpers/LICENSE).
```
Comment (ISSOtm) [inline README.md:65]:
```suggestion
- The code related to the i18n support is originally from Google's [Comprehensive Rust](https://github.com/google/comprehensive-rust) and is released under the [Apache License 2.0](https://github.com/gbdev/gb-asm-tutorial/blob/master/i18n-helpers/LICENSE).
```

File: book.toml [hunk 6]
```diff
@@ -8,8 +8,9 @@ src = "src"
 [build]
 create-missing = true # This is kept for convenience, but CI sets it to false
 use-default-preprocessors = false
+extra-watch-dirs = ["po"]
 
-[preprocessor.links]
+[preprocessor.gettext]
```
Comment (ISSOtm) [inline book.toml:13]:
```suggestion
[preprocessor.links]

[preprocessor.gettext]
```
The built-in `links` preproc is required to expand `{{#include}}`.

That said, this presents an interesting question: in which order should these two run? I.e. should the `include`d text be localised? I'd say no, and then we should add a `before = ["links"]` line to the `[preprocessor.gettext]` table.

File: po/it.po [hunk 7]
```diff
@@ -0,0 +1,5605 @@
+msgid ""
+msgstr ""
+"Project-Id-Version: GB ASM Tutorial\n"
+"POT-Creation-Date: \n"
+"PO-Revision-Date: 2023-03-02 23:18+0100\n"
+"Last-Translator: avivace4@gmail.com\n"
+"Language-Team: Italian <tp@lists.linux.it>\n"
```
Comment (ISSOtm) [inline po/it.po:7]: Um, is that right?

File: theme/index.hbs [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+                                aria-controls="language-list">
+                            <i class="fa fa-globe"></i>
+                        </button>
+                        <ul id="language-list" class="theme-popup" aria-label="Languages" role="menu">
+                          <li role="none"><button role="menuitem" class="theme">
+                              <a id="en">English</a>
+                          </button></li>
+                          <li role="none"><button role="menuitem" class="theme">
+                              <a id="it">Italian</a>
+                          </button></li>
```
Comment (ISSOtm) [inline theme/index.hbs:151]: We should probably mention adding this in the "initialising a new translation" section.

File: TRANSLATING.md [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+* `mdbook-xgettext`: This program extracts the English text. It is an mdbook
+  renderer.
+* `mdbook-gettext`: This program translates the book into a target language. It
+  is an mdbook preprocessor.
+
+Install both helpers with the following command from the root of the course:
+
+```shell
+$ cargo install --path i18n-helpers
+```
```
Comment (ISSOtm) [inline TRANSLATING.md:37]: I'm not a total fan of making people install software globally on their machine, especially due to the possibility of version conflicts. Would it be possible to use a `command` entry instead, like for the other two custom programs we already have? `cargo run -p i18n-helpers --bin mdbook-gettext --locked --release --` should do the trick, I think.

(If we do go through with that, then this paragraph and the corresponding step in CI should be removed.)

**Review by ISSOtm — CHANGES_REQUESTED** [review]
Thank you very much for your help setting this up! I only have suggestions about some details; the overall process is very good!

File: README.md (hunk 4 above)
Comment (avivace) [reply README.md:26]: I moved it to its own chapter. It will also have a paragraph about contributing directly on Crowdin.

File: README.md (hunk 3 above)
Comment (avivace) [reply README.md]: Done

File: book.toml (hunk 6 above)
Comment (avivace) [reply book.toml:15]: Agree, I changed as suggested

File: theme/index.hbs (hunk 8 above)
Comment (avivace) [reply theme/index.hbs:151]: Done

File: po/it.po (hunk 7 above)
Comment (avivace) [reply po/it.po:7]: I don't think that line is mandatory, I removed it

File: TRANSLATING.md (hunk 9 above)
Comment (avivace) [reply TRANSLATING.md:37]: Done!

File: README.md [hunk 10]
```diff
@@ -10,17 +10,26 @@ To deploy gb-asm-tutorial locally:
...
 3. The HTML files are in `book/custom/`.
 
   ⚠️ `book/html/` contains only partially processed files.
   This folder is what gets served when running `mdbook serve`, so you may see some custom markup missing if using that.
   As a workaround, you can manually open the files in the `book/custom/` folder in your browser, they just won't auto-refresh on changes.
 
+## Translating
+
+To add a translation or contribute on an existing one, please see [TRANSLATING](TRANSLATING.md).
+
```
Comment (ISSOtm) [inline README.md:26]: Shouldn't this be above the "contributing" header, so it's more prominent?

File: TRANSLATING.md [hunk 11]
```diff
@@ -0,0 +1,120 @@
+# Translating the GB ASM Tutorial
+
+We would love to have your help with translating the tutorial into other
+languages! We use the [Gettext] system for translations. This means that you
+don't modify the Markdown files directly: instead you modify `.po` files in a
+`po/` directory. The `.po` files are small text-based translation databases.
+
```
Comment (ISSOtm) [inline TRANSLATING.md:7]: I think we should mention Crowdin here. It should have less friction than the whole PR workflow, and so be more successful.

File: theme/index.hbs [hunk 12]
```diff
@@ -0,0 +1,362 @@
...
+                                aria-haspopup="true" aria-expanded="false"
+                                aria-controls="language-list">
+                            <i class="fa fa-globe"></i>
+                        </button>
+                        <ul id="language-list" class="theme-popup" aria-label="Languages" role="menu">
+                          <li role="none"><button role="menuitem" class="theme">
+                              <a id="en">English</a>
+                          </button></li>
+                          <li role="none"><button role="menuitem" class="theme">
+                              <a id="it">Italian</a>
```
Comment (ISSOtm) [inline theme/index.hbs:150]: Hey, actually, what does clicking this link do?

**Review by ISSOtm — CHANGES_REQUESTED** [review]

File: TRANSLATING.md (hunk 11 above)
Comment (avivace) [reply TRANSLATING.md:7]: I agree  but I suggest doing this in two phases. The crowdin workflow is still not tested yet and I'd feel more comfortable in starting to merge the translation system first then see if crowdin works as a collaborative frontend.

File: theme/index.hbs (hunk 12 above)
Comment (avivace) [reply theme/index.hbs:150]: It links to the italian version of the current page. 

See demo over: https://avivace.github.io/gb-asm-tutorial/

Comment (ISSOtm) [reply theme/index.hbs:150]: The lack of `href` really tripped me; apparently that's set up via JS, which I really don't like (why not static in the first place?) but I guess it has to do for now.

File: TRANSLATING.md (hunk 11 above)
Comment (ISSOtm) [reply TRANSLATING.md:7]: Fine by me; then this PR is basically complete.

File: README.md (hunk 10 above)
Comment (avivace) [reply README.md]: done

File: theme/index.hbs (hunk 12 above)
Comment (avivace) [reply theme/index.hbs:156]: I guess that would require another preprocessor populating the submenu? Feel free to open an issue about it if you feel like it's necessary

**Review by ISSOtm — APPROVED** [review]
That's step 1 done :)

**Comment by mgeisler** [comment]
Hi folks! Thanks for trying out the i18-helpers, I hope they're useful?

I just wanted to let you know that I've released them as a stand-alone crate now: https://crates.io/crates/mdbook-i18n-helpers. That should let you clean up the code here a bit :slightly_smiling_face: 

I noticed you use `cargo run` directly. With a downloaded crate, you should be able to replace that with `cargo install --root tmp mdbook-i18n-helpers` and then refer to `tmp/bin/mdbook-xgettext` and `tmp/bin/mdbook-gettext` in your `book.toml`.

Please open issues in the repository for any problems you run into. Thanks!

**Comment by avivace** [comment]
> Hi folks! Thanks for trying out the i18-helpers, I hope they're useful?
> 
> I just wanted to let you know that I've released them as a stand-alone crate now: [crates.io/crates/mdbook-i18n-helpers](https://crates.io/crates/mdbook-i18n-helpers). That should let you clean up the code here a bit slightly_smiling_face
> 
> I noticed you use `cargo run` directly. With a downloaded crate, you should be able to replace that with `cargo install --root tmp mdbook-i18n-helpers` and then refer to `tmp/bin/mdbook-xgettext` and `tmp/bin/mdbook-gettext` in your `book.toml`.
> 
> Please open issues in the repository for any problems you run into. Thanks!

Wow, thanks a lot @mgeisler ! I'll open another issue with the points you mentioned so we can take a look at this later. Thanks again for the amazing translation system, it's working very nice for us!

**Comment by mgeisler** [comment]
> Thanks again for the amazing translation system, it's working very nice for us!

Thanks for the feedback, I'm very happy it works for you!

**Comment by mgeisler** [comment]
Hey @avivace, I wanted to let you know about https://github.com/google/mdbook-i18n-helpers/pull/46.

The current `main` branch of mdbook-i18n-helpers has a new text extraction system: it will turn `# Some heading` in the Markdown file into just `"Some heading"` in the PO file. Paragraphs are un-wrapped when extracted. Lists are now extracted on a per-item basis and tables are extracted on a cell-by-cell basis. This removes most of the Markdown syntax and makes the job easier for translators. It also allows you to freely reformat the Markdown files without invalidating the existing translations.

For this to work smoothly with existing translations, I've written a conversion tool: https://github.com/google/mdbook-i18n-helpers/pull/46. When that is merged, I will make a new release. I hope you will be able to migrate to it as well.

Eventually, I want to release a GitHub action for all this: https://github.com/google/mdbook-i18n-helpers/issues/48. That should make your life easier since you can export the complexity back to the mdbook-i18n-helpers repository :smile:

**Comment by mgeisler** [comment]
I added a link to this projects in the mdbook-i18n-helpers README: https://github.com/google/mdbook-i18n-helpers/pull/49. I hope you don't mind?

**Comment by mgeisler** [comment]
> This sounds very interesting and cleaning up the strings from formatting is definitely appreciated. Feel free to ping me when the migration tool is ready!

The converter tool is part of the new 0.2 release which I made last week.

I've created PRs for Comprehensive Rust today to update the translations. As an example, the [Brazilian Portuguese page on scalar types](https://google.github.io/comprehensive-rust/pt-BR/basic-syntax/scalar-types.html) is using the new system — the table you see there is translated cell-by-cell. If we change one of them in the English original, the others will remain translated.

**Pull request merged**